src/Host/Client/Impl/Definitions/Message.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using Microsoft.Common.Core;
+using Newtonsoft.Json.Linq;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Concurrent;
+using System.Text;
+using Newtonsoft.Json;
+using static System.FormattableString;
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.R.Host.Client {
+    public class Message {
+        public readonly ulong Id;
+        public readonly ulong RequestId;
+        public readonly string Name;
+
+        public readonly JArray Json;
+        public readonly byte[] Blob;
+
+        public bool IsRequest => RequestId == ulong.MaxValue;
+
+        public bool IsNotification => RequestId == 0;
+
+        public bool IsResponse => !IsRequest && !IsNotification;
+
+        public int ArgumentCount => Json.Count;
+
+        public Message(ulong id, ulong requestId, string name, JArray json, byte[] blob = null) {
+            Id = id;
+            RequestId = requestId;
+            Name = name;
+            Json = json;
+            Blob = blob ?? new byte[0];
+        }
+
+        public Message(byte[] data) {
+            try {
+                int offset = 0;
+                Id = BitConverter.ToUInt64(data, offset);
+                offset += sizeof(ulong);
+
+                RequestId = BitConverter.ToUInt64(data, offset);
+                offset += sizeof(ulong);
+
+                int term = Array.IndexOf<byte>(data, 0, offset);
+                if (term < 0) {
+                    throw new IndexOutOfRangeException();
+                }
+                Name = Encoding.UTF8.GetString(data, offset, term - offset);
```

Comment:
I started down that path, but there were more extensive changes on the host side for this (and other things) that I didn't want to get into, given the tight schedule.

For now, I think this will do even if we do need super-fast name checking in the broker, because the name is positioned at a fixed location inside the message, and most interesting messages have short names. So one can cast the pointer to name to a pointer to some appropriate integer type and compare (e.g. if you need to spot evals, you'd cast to int16 and compare against `?=`).